apps/oracles/hyper-borrow-rates/src/providers/euler.rs

```diff
@@ -0,0 +1,160 @@
+
+    let ret = euler::EulerUtilsLens::getAPYsCall::abi_decode_returns(&raw)?;
+
+    Ok(ret.borrowAPY)
+}
+
+pub async fn fetch_borrow_rates_from_euler(
+    feeds_config: Option<&Vec<FeedConfig>>,
+) -> Result<RatesPerFeed> {
+    let mut borrow_rates: RatesPerFeed = RatesPerFeed::new();
+
+    let Some(feeds_config) = feeds_config else {
+        return Ok(borrow_rates);
+    };
+
+    let feeds_by_network = group_feeds_by_network(feeds_config);
+
+    borrow_rates.extend(
+        fetch_borrow_rates_on_ethereum_mainnet(
+            feeds_by_network
+                .get(&SupportedNetworks::EthereumMainnet)
+                .unwrap_or(&Vec::new()),
+        )
+        .await?,
+    );
+    Ok(borrow_rates)
+}
+
+async fn fetch_borrow_rates_on_ethereum_mainnet(feeds: &Vec<FeedConfig>) -> Result<RatesPerFeed> {
+    let mut borrow_rates: RatesPerFeed = RatesPerFeed::new();
+
+    let rpc_url = Url::parse(ETHEREUM_MAINNET_RPC_URL)?;
+    let provider = Arc::new(ProviderBuilder::new().connect_http(rpc_url.clone()));
+
+    let mut futures = FuturesUnordered::new();
+
+    for feed in feeds {
+        let utils_lens_address = feed.arguments.utils_lens_address.clone().unwrap();
+        let vault_address = feed.arguments.vault_address.clone().unwrap();
+
+        let feed_id = feed.feed_id.clone();
+        let asset = feed.pair.base.clone();
+        let provider = provider.clone();
+        let fut = async move {
+            let result = fetch_borrow_rate(&provider, utils_lens_address, vault_address).await;
+            (feed_id, asset, result)
+        };
+
+        futures.push(fut);
+    }
+
+    while let Some((feed_id, asset, result)) = futures.next().await {
+        match result {
+            Result::Ok(rate) => {
+                borrow_rates.insert(
+                    feed_id,
+                    BorrowRateInfo {
+                        asset: asset,
+                        underlying_asset: None,
+                        borrow_rate: apr_continuous(rate),
+                    },
+                );
+            }
+            Result::Err(err) => {
+                eprintln!("Error fetching borrow rate for feed {}: {}", feed_id, err);
+            }
+        }
+    }
+
+    Ok(borrow_rates)
+}
+
+pub fn group_feeds_by_network(
+    feeds_config: &Vec<FeedConfig>,
+) -> HashMap<SupportedNetworks, Vec<FeedConfig>> {
+    let mut grouped: HashMap<SupportedNetworks, Vec<FeedConfig>> = HashMap::new();
+
+    for feed in feeds_config {
+        // check that addresses are in config
+        if feed.arguments.network.is_none()
+            || feed.arguments.utils_lens_address.is_none()
+            || feed.arguments.vault_address.is_none()
+        {
+            eprintln!(
+                "Feed {} missing network or required addresses. Skipping.",
+                feed.feed_id
+            );
+            continue;
+        }
+
+        let network = feed.arguments.network.clone().unwrap();
```

Comment:
Don't just check, [parse](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)

```suggestion
        // Addresses must be in the config
        let network = match (
            feed.arguments.network,
            feed.arguments.utils_lens_address,
            feed.arguments.vault_address,
        ) {
            (Some(network), Some(_), Some(_)) => network,
            _ => eprintln!(
                "Feed {} missing network or required addresses. Skipping.",
                feed.feed_id
            ),
        };
```

Reply:
This also begs the question, should be not enhance the type of objects inside the `Vec<FeedConfig>`? Since we validate here that all three of `network`, `utils_lens_address` and `vault_address`, why not pass their real values forward? Otherwise we'd need to `.unwrap()` later or make never-matching error cases to be type-safe, even tho we (but not the compiler) *know* that those three fields will never be `None`.

Reply:
Will handle in next PR